test/DotNetty.Transport.Tests.Performance/Sockets/TcpSocketChannelPerfSpec.cs

```diff
@@ -0,0 +1,122 @@
+        {
+            return new LengthFieldPrepender(4, false);
+        }
+
+        protected virtual IChannelHandler GetDecoder()
+        {
+            return new LengthFieldBasedFrameDecoder(20, 0, 4, 0, 4);
+        }
+
+        [PerfSetup]
+        public void SetUp(BenchmarkContext context)
+        {
+            this.ClientGroup = new MultithreadEventLoopGroup(1);
+            this.ServerGroup = new MultithreadEventLoopGroup(1);
+            this.WorkerGroup = new MultithreadEventLoopGroup();
+
+            Encoding iso = Encoding.GetEncoding("ISO-8859-1");
+            this.message = iso.GetBytes("ABC");
+
+            this._inboundThroughputCounter = context.GetCounter(InboundThroughputCounterName);
+            this._outboundThroughputCounter = context.GetCounter(OutboundThroughputCounterName);
+            var counterHandler = new CounterHandlerInbound(this._inboundThroughputCounter);
+            this._signal = new ManualResetEventSlimReadFinishedSignal(this.ResetEvent);
+
+            ServerBootstrap sb = new ServerBootstrap().Group(this.ServerGroup, this.WorkerGroup).Channel<TcpServerSocketChannel>()
+                .ChildHandler(new ActionChannelInitializer<TcpSocketChannel>(channel => { channel.Pipeline.AddLast(this.GetEncoder()).AddLast(this.GetDecoder()).AddLast(counterHandler).AddLast(new CounterHandlerOutbound(this._outboundThroughputCounter)).AddLast(new ReadFinishedHandler(this._signal, WriteCount)); }));
+
+            Bootstrap cb = new Bootstrap().Group(this.ClientGroup).Channel<TcpSocketChannel>().Handler(new ActionChannelInitializer<TcpSocketChannel>(
+                channel =>
+                {
+                    channel.Pipeline.AddLast(this.GetEncoder()).AddLast(this.GetDecoder()).AddLast(counterHandler)
+                        .AddLast(new CounterHandlerOutbound(this._outboundThroughputCounter));
+                }));
+
+            // start server
+            this._serverChannel = sb.BindAsync(TEST_ADDRESS).Result;
+
+            // connect to server
+            this._clientChannel = cb.ConnectAsync(this._serverChannel.LocalAddress).Result;
+        }
+
+        [PerfBenchmark(Description = "Measures how quickly and with how much GC overhead a TcpSocketChannel --> TcpServerSocketChannel connection can decode / encode realistic messages",
+            NumberOfIterations = 13, RunMode = RunMode.Iterations)]
+        [CounterMeasurement(InboundThroughputCounterName)]
+        [CounterMeasurement(OutboundThroughputCounterName)]
+        [GcMeasurement(GcMetric.TotalCollections, GcGeneration.AllGc)]
+        [MemoryMeasurement(MemoryMetric.TotalBytesAllocated)]
+        public void TcpChannel_Duplex_Throughput(BenchmarkContext context)
+        {
+            for (int i = 0; i < WriteCount; i++)
+            {
+                this._clientChannel.WriteAsync(Unpooled.WrappedBuffer(this.message).Retain());
+                if (i % 10 == 0) // flush every 10 writes
```

Comment:
I'd have to add them as separate specs, which extend the runtime - but I did notice a big improvement between 10 and 100.

However, if we're also pumping the benchmark up to 1 million messages I can also cut down on the run interval.